File: apps/api/app/schemas/discovery.py
"""Web price-discovery schemas.

These describe *estimated* sellers/prices gathered from the open web (not
verified invoices). Every field except ``name`` is optional because a given
listing may only surface a name + a link.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class DiscoveredSeller(BaseModel):
    name: str
    # Raw price string as it appeared, kept verbatim for transparency.
    price_text: str | None = None
    # Best-effort structured pack price.
    price_cents: int | None = None
    pack_desc: str | None = None
    pack_qty: float | None = None
    pack_unit: str | None = None  # one of PackUnit values when known
    # Normalized (server-computed from the fields above) for comparison.
    unit_price_cents: float | None = None
    base_unit: str | None = None
    # Provenance.
    url: str | None = None
    location: str | None = None
    distance_note: str | None = None  # free-text proximity ("~18 mi, Redding")
    snippet: str | None = None


class DiscoverResponse(BaseModel):
    configured: bool
    query: str
    sellers: list[DiscoveredSeller] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)
    disclaimer: str = (
        "Estimated from public web sources — verify pricing and availability with "
        "the seller before ordering."
    )

File: apps/web/src/lib/discovery.ts
/** Web price-discovery types + client. Results are ESTIMATES from the open web. */

import { apiGet } from "@/lib/api";

export interface DiscoveredSeller {
  name: string;
  price_text: string | null;
  price_cents: number | null;
  pack_desc: string | null;
  pack_qty: number | null;
  pack_unit: string | null;
  unit_price_cents: number | null;
  base_unit: string | null;
  url: string | null;
  location: string | null;
  distance_note: string | null;
  snippet: string | null;
}

export interface DiscoverResponse {
  configured: boolean;
  query: string;
  sellers: DiscoveredSeller[];
  notes: string[];
  disclaimer: string;
}

export async function discoverPrices(
  ingredientId: string,
  opts: { radiusMiles: number; location?: string },
): Promise<DiscoverResponse | null> {
  const params = new URLSearchParams({ radius_miles: String(opts.radiusMiles) });
  if (opts.location && opts.location.trim().length > 0) {
    params.set("location", opts.location.trim());
  }
  const res = await apiGet<DiscoverResponse>(
    `/api/v1/ingredients/${ingredientId}/discover-prices?${params.toString()}`,
  );
  return res.ok ? res.data : null;
}
